Initial version of a C++ backend

parno opened the pull request:
This is a preliminary C++ backend for Dafny.  As discussed in Issue #563, it has a number of limitations (also documented in `Test/c++/LIMITATIONS.md`).  As a result of these limitations, the C++ compiler comes with its own set of systems-oriented compiler tests.  At the moment, they are driven by a custom `Makefile` and Python script, since (a) I'm uncertain how to get `lit` to compile C++ code using the ambient C++ compiler, and (b) it seems like Dafny will be moving to xUnit soon (#680).

It may be easier to review the two commits separately, as one has all of the C++ tests, while the other is the actual code change.  This PR makes one additional change to the generic `Compiler.cs` (namely expanding the `CreateGetter` interface with additional arguments) to cope with a recent change that broke some C++ functionality.

robin-aws commented:
Exciting! :) I'm working now on finishing off #680 and will prioritize getting it in first. I'll have a look at what will be necessary to support the new tests in that framework - I'm sure we could work out how to get lit to do it but that seems like wasted effort.

robin-aws commented:
I think the best way to support the tests is to implement `CompileTargetProgram` and `RunTargetProgram` in `CppCompiler`. That will mean a command like `dafny /compile:3 /compileTarget:cpp hello.dfy` (which is all lit is doing for the most part) will generate the C++ code from Dafny, compile the generated code, and then run the compiled binary. That means you won't need C++-specific tests either - all the ones you include can also be run for other target languages for completeness. We can also mark all the existing test cases that DON'T work for C++ yet as known issues once #680 lands.

You should just refer to `g++` directly in `CompileTargetProgram`, just as the java compiler refers directly to whatever `javac` is on the path. It's not very flexible, but consistent with the other compilers. For any more complicated contexts, I'd expect users to just use `dafny` to generate the target code and then rely on target language-specific tooling for the rest of the build process. The dafny.msbuild plugin uses this approach for example: https://github.com/dafny-lang/dafny.msbuild/blob/master/Source/dafny.msbuild/build/dafny.msbuild.targets#L19

RustanLeino commented on Test/c++/LIMITATIONS.md:
```diff
@@ -0,0 +1,18 @@
+The C++ backend was written assuming that it would primarily support writing
+C/C++ style code in Dafny, which leads to some limitations in the current
+implementation.
```
This file is better placed in the docs/Compilation folder.

parno replied:
Ah, yes, that's a much better place.

RustanLeino commented on Test/c++/LIMITATIONS.md:
```diff
@@ -0,0 +1,18 @@
+C/C++ style code in Dafny, which leads to some limitations in the current
+implementation.
+
+- We do not support BigIntegers, so do not use `int`, or raw instances of
+  `arr.Length`, or sequence length, etc. in executable code.  You can however,
+  use `arr.Length as uint64` if you can prove your array is an appropriate
+  size.  The compiler will report inappropriate integer use.
```
Is this a change from a few months ago when (I thought) `int` was compiled as some fixed-width C++ integer type? Or is the new compiler compiling `int` as a fixed-length integer, but this happens to be okay as long as the Dafny program doesn't use `int` (other than temporarily in expressions like `arr.Length as uint64`)?

Does the compiler generate an error if it would have occasion to generate code that would need a `BigInteger`?

parno replied:
No, this is how we've consistently used the C++ backend.  We expect the programmer to avoid the use of mathematical integers in their code and instead to use subset types that fit into native types.  That ensures our compilation remains sound.

At the moment, if the C++ backend encounters a BigInteger, it emits a warning and C++ code for a non-existent BigInteger implementation.  We found it useful to warn instead of error, since it gives you a sense of where all of the `int` issues are, rather than just showing you one at a time.

RustanLeino replied:
Sounds good. That suggests it should be possible to link up the code with a C++ BigInteger library and then run the Dafny-to-C++ compiler on the rest of Dafny's compiler tests.

parno replied:
Indeed.  I wasn't sure if Dafny was ready to take a long-term dependency on Boost, GMP (GPL license), or other related library, since C++ doesn't come with its own built-in library, but it's something that could be pursued in the future.  Perhaps if you find an ambitious Amazon intern, they could implement a library in Dafny itself :)

robin-aws replied:
I've been insisting to Rustan for a while now that we need to make the compilation process pluggable, so that user codebases can CHOOSE to take such dependencies as they see fit and compile built-in Dafny concepts to whatever implementations they prefer. :) Figuring out the exact plugin interface for that will be tricky, however.

RustanLeino commented on Test/c++/arrays.dfy:
```diff
@@ -0,0 +1,96 @@
+newtype{:nativeType "uint"} uint32 = i:int | 0 <= i < 0x100000000
+
+method returnANullArray() returns (a: array?<uint32>)
+ensures a == null
```
indent

RustanLeino commented on Test/c++/arrays.dfy:
```diff
@@ -0,0 +1,96 @@
+}
+
+method returnANonNullArray() returns (a: array?<uint32>)
+ensures a != null
+ensures a.Length == 5
```
indent

RustanLeino commented on Test/c++/arrays.dfy:
```diff
@@ -0,0 +1,96 @@
+newtype{:nativeType "uint"} uint32 = i:int | 0 <= i < 0x100000000
```
The Dafny implementation has hooks for assigning native types automatically. If they are used, you shouldn't need the explicit `:nativeType` annotation here (or in the other test files).

parno replied:
Ah, interesting, I hadn't realized that.  I'll experiment with removing the annotations.

RustanLeino commented on Test/c++/arrays.dfy:
```diff
@@ -0,0 +1,96 @@
+}
+
+method PrintArray<A>(a:array?<A>, len:uint32) 
+  requires a != null ==> len as int == a.Length;
```
remove the final semi-colon

RustanLeino commented on Test/c++/class.dfy:
```diff
@@ -0,0 +1,177 @@
+  var ar: array<uint32>
+
+  constructor (x: uint32) 
+    requires x < 100;
+    ensures this.a < 300;
+    ensures this.b < 300;
+    ensures fresh(this.ar);
```
remove semi-colons

RustanLeino commented on Test/c++/class.dfy:
```diff
@@ -0,0 +1,177 @@
+  static method N() returns (r: uint32) { return 70; }
+
+  method stuffWithAr()
+  modifies this
+  modifies this.ar
```
indent

RustanLeino commented on Test/c++/class.dfy:
```diff
@@ -0,0 +1,177 @@
+  a.y := 15;
+
+  if a == a {
+      print "EqualityReflexive: This is expected\n";
+  } else {
+      print "EqualityReflexive: This is unexpected\n";
+      assert false;
```
Here and throughout, use a 2-space indent

RustanLeino commented on Test/c++/datatypes.dfy:
```diff
@@ -0,0 +1,145 @@
+}
+
+method Callee(e:Example2) {
+    match e {
+        case Ex2a(u) => print "Case A: ", u, "\n";
+        case Ex2b(b) => print "Case B: ", b, "\n";
+    }
```
Here and below, use a 2-space indent

RustanLeino commented on Source/Dafny/Compiler-cpp.cs:
```diff
@@ -0,0 +1,2338 @@
+    // Dedicated hash-function definitions for each type
+    private TargetWriter hashWr = null;
+
+    // Shadowing variables in Compiler.cs
+    new string DafnySetClass = "DafnySet";
+    new string DafnyMultiSetClass = "DafnyMultiset";
+    new string DafnySeqClass = "DafnySequence";
+    new string DafnyMapClass = "DafnyMap";
```
Override them instead

parno replied:
I don't think C# will let me override a field, will it?  Or are you suggesting upgrading the existing fields in `Compiler.cs` to properties so that they can be overridden?

RustanLeino replied:
Yes, I was thinking converting the original types to be virtual property getters. I just suggested the same thing in https://github.com/dafny-lang/dafny/pull/696, where this also came up. We can adjust these once these PRs are both merged into master.

parno replied:
Okay, sounds good to me.

RustanLeino commented on Source/Dafny/Compiler-cpp.cs:
```diff
@@ -0,0 +1,2338 @@
+    private List<string> classDefaults;
+    
+    /*
+     * Unlike other Dafny and Dafny's other backends, C++ cares about
+     * the order in which types are declared.  To make this more likely
+     * to succeed, we emit type information as gradually as possible
+     * in hopes that definitions are in place when needed.
```
Would it help to emit stub declarations for types into .h files? The Java and Go back-ends have facilities to write some multiple files.

parno replied:
The C++ backend does start by emitting (all of) the type declarations as stubs at the top of the single `.h` file it produces.  That helps, but unfortunately there are still some complexities for things like emitting the class definition if one class contains another (or a class contains a datatype), where C++ will object if it doesn't know the size of the type.

RustanLeino commented on Source/Dafny/Compiler-cpp.cs:
```diff
@@ -0,0 +1,2338 @@
+        bool isHandle = true;
+        if (cl != null && Attributes.ContainsBool(cl.Attributes, "handle", ref isHandle) && isHandle) {
+          return "ulong";
+        } else if (DafnyOptions.O.IronDafny &&
```
This question is out of place here, but perhaps you know the answer: Is `DafnyOption.O.IronDafny` used by anyone anymore? If it's not needed for C++, you can remove the case here (and the same for the other back ends).

parno replied:
Jay, Jon, and I are no longer using it.  I'm waiting to hear back from Manos.  If he's not, then I think it would be safe to remove at this point.

parno commented:
> I think the best way to support the tests is to implement `CompileTargetProgram` and `RunTargetProgram` in `CppCompiler`. That will mean a command like `dafny /compile:3 /compileTarget:cpp hello.dfy` (which is all lit is doing for the most part) will generate the C++ code from Dafny, compile the generated code, and then run the compiled binary. That means you won't need C++-specific tests either - all the ones you include can also be run for other target languages for completeness. We can also mark all the existing test cases that DON'T work for C++ yet as known issues once #680 lands.
> 
> You should just refer to `g++` directly in `CompileTargetProgram`, just as the java compiler refers directly to whatever `javac` is on the path. It's not very flexible, but consistent with the other compilers. For any more complicated contexts, I'd expect users to just use `dafny` to generate the target code and then rely on target language-specific tooling for the rest of the build process. The dafny.msbuild plugin uses this approach for example: https://github.com/dafny-lang/dafny.msbuild/blob/master/Source/dafny.msbuild/build/dafny.msbuild.targets#L19

@robin-aws Good idea.  That was simpler than I expected, so I went ahead and implemented those two routines and updated the C++ tests with basic lit headers for C++.  I haven't yet tried them on the other languages.

parno commented:
@robin-aws Looking at the logs for the checks above, they appear to be for my previous commit 8d12995b3a7, rather than my most recent commit c489a8a, even when I click the button to re-run the tests.  Do you know how I can kick off a run for the most recent commit?

```
/usr/bin/git checkout --progress --force refs/remotes/pull/702/merge
...
HEAD is now at 4b79bfb Merge 8d12995b3a79efda4f437419d2cb93c695e7580f into c0cac2ade8ed38210c4595c5eaa410ca68b49bc5
```

robin-aws commented:
@parno Hmm, I agree it doesn't seem to be picking up the right commits. Looking into it.

robin-aws commented:
@parno Can you try pushing an empty commit on the cpp-backend branch? I don't have permissions to.

parno commented:
@robin-aws Thanks, the empty commit seems to have kicked things off with the right commit this time.

I'm curious if you have any thoughts on how to make the compilation tests more predictable/stable.  At the moment, I'm invoking `g++` with `-Wall`, since it seems like a good idea to have as many safety checks as possible on the (unverified) compiler.  However, this leads to mismatched warnings, e.g., on `Test/c++/datatypes.dfy`, as shown below.  There's a Dafny match statement that produces an unused variable (on all compiler backends), and it appears that the CI machine's `g++` formats the warning slightly differently than my `g++` compiler.  Obviously, I can change the `.expect` file to match the CI machine's output, but I worry that this will still be flaky on other people's machines.  At the moment, Dafny doesn't support passing additional compiler arguments (to, e.g., suppress the warning for this particular test).

 ```
+ In static member function 'static _module::uint32 _module::__default::IntListLen(_module::IntList)':
 339:16: warning: unused variable '_131_hd' [-Wunused-variable]
-        uint32 _131_hd = _128___mcc_h0;
-               ^
-1 warning generated.
+         uint32 _131_hd = _128___mcc_h0;
+                ^~~~~~~
```

RustanLeino commented:
Is it possible to use `-Wall`, but then `+Wunused_variables` or something like that? The other compiler back-ends silence various compiler warnings: the C# back-end does `/nowarn:0219 /nowarn:0168` to suppress warnings for unused variables (it uses two warning numbers, because of differences in C# compilers, the Java back-end sprinkles `@SuppressWarnings({"unchecked", "deprecation"})` here and there, and the Go back-end emits things like `_ = x` after the declaration of every variable to avoid Go's unused-variable errors.

Also, it seems the `Test/c++/bit-vectors.dfy` test is failing somehow.

robin-aws commented:
@parno Good question! The easiest and most consistent solution would be to not enable those warnings; the other compilers turn off several similar warnings that the Dafny-compiled source trips already (for example: https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Compiler-Csharp.cs#L2415-L2422) I've carried over that configuration over into dafny.msbuild as well, although I don't love the fact that this disables those warnings for manually-written C# code in your project. The longer-term fix should be separate compilation, but that still has a few blockers in the way: #501. 

Having said that, I'm also running into the need for .expect files to specify a partial match of some kind as part of #680: attempting to compile iterators in Java, for example, currently prints out a stack trace that is similarly unpredictable. However, my preference is to print less information unless additional debugging is enabled, so that the output becomes predictable. I'm working on converting all the tests now and will address this if there are a lot of known issues with unpredictable output.

parno commented:
Thanks @robin-aws and @RustanLeino.  I hadn't realized the other back-ends were actively suppressing warnings.  I updated the C++ backend to suppress unused variable warnings and tweaked my tests to avoid other variants thereof, so I think everything should be in a passing state, assuming nuget cooperates this time.
